### ISSUE TITLE
PR #3397 somehow merged without updated snapshots (this repo only)

### ISSUE THREAD

**Issue by giseburt** [projen/projen#3410]
Note: This is about _this_ project specifically, not configuring other projects.

## Theory: Default `jestConfig.updateSnapshot` allowing test failures to pass silently

In the `.projenrc.js` of the `projen` project itself, `jestOptions.jestConfig.updateSnapshot` is not specified, and due to historical reasons (it appears, looking at closed issues) this defaults to `ALWAYS`.

What happens is that any tests that **should** fail because the snapshot changed doesn't, instead the snapshot is silently "fixed".

This causes a number of bad things at once:

- The test passed, when it shouldn't have, masking the issue the test it supposed to catch
- We don't see _which_ issue caused the snapshot mismatch (it can be determined, but with more effort than it should be)

For this reason, I'm suggesting we change the `.projenrc.js` file *of this repo* to:

```js
const project = new cdk.JsiiProject({
  // ...
  jestOptions: {
    // ...
    jestConfig: {
      // ...
      updateSnapshot: NEVER, // ← add this
    },
  },
```

There is a mechanism in place to detect if file changes would occur such as when a snapshot changes. This *helps* but recently allowed a code change that should have caused a test failure (in a PR that I submitted, #3397) that ended up causing the build to fail in another PR I submitted ([here](https://github.com/projen/projen/pull/3409/commits/68d67dfda28755b0c3554d6d3c820d026879953e)). I'm not sure why it didn't catch it, but more importantly it should have been caught on the test build of the first PR and adjusted there.

**Comment by mrgrain** [comment]
I think the issue is more subtle than that.

This PR introduced the snapshots: https://github.com/projen/projen/pull/3406
This is the build that got the PR merged: https://github.com/projen/projen/actions/runs/8079427878/job/22073814706#step:5:1014

As you can see, there is no mention of updated snapshots.

---

I suspect we have test cases that install the latest version of projen, rather than using the locally build version.

**Comment by gmeligio** [comment]
I tried to narrow down when the snapshot got outdated by checking out older commits from the main branch, and I could reproduce it in all of them before the fix without finding the first commit. I deleted the repo multiple times to check there were no caching issues and still got the same result. The file `test/__snapshots__/new.test.ts.snap` is modified during all my tests for commits older than https://github.com/projen/projen/commit/75072c1ed699dd0e02ecd0627e72c4ff907c9728.

What I did was:
1. git checkout <git-ref-sha>  
1. `yarn && rm -rf lib && yarn projen && yarn test test/new.test.ts`
1. git diff

I think the snapshot configuration, generation and verification can be discarded from the possible causes. Those looked fin when I tested them. It looks more like an external factor in `test/new.test.ts` that caused the change in the snapshot. It could be what @mrgrain said about some tests not using the local projen.

I haven't been able to identify what is the cause but have narrowed down the time window for the change in that external dependency, which happened between these two jobs:
1. Successful build in PR [#3406](https://github.com/projen/projen/pull/3406): https://github.com/projen/projen/actions/runs/8079427878?pr=3406
2. Failed release in main: https://github.com/projen/projen/actions/runs/8079532064

**Comment by mrgrain** [comment]
Looking at the snapshot change, I think this was actually caused by https://github.com/projen/projen/pull/3397 that went in just before your change @gmeligio

**Comment by giseburt** [comment]
I believe it was caused by #3397, and should have been caught and fixed there. But the tests passed without error.

Snapshot updating is only supposed to happen while making or modifying tests. If the auto update feature were disabled I’d have seen the test failure before I pushed the code and would have been able to evaluate if that was an acceptable change in results.

**Comment by mrgrain** [comment]
Apologies, I missed that I'm your original investigation. So we agree on the change that introduced it. 

I still don't think the snapshot update is the cause here. The builds on #123 don't show any updated snapshots either. For fail on change to have helped, the tests would have to had run with different results locally and on CI. That means you would have seen snapshot changes locally but decided to not commit them. Is that what happend? And even then, why does the CI behave differently? Would the CI then have failed due to it producing the outdated template?

---

The snapshot mutation in CI exists so maintainers like me get the benefit of self-updating snapshots. Ideally you'd would also work for forks, but we haven't figured that out yet. Also we use this everywhere and it's not been an issue except in projen (afair this is the second time). Hence my thinking this is an issue with a test case.

**Comment by gmeligio** [comment]
I wanted to investigate more today, but couldn't find the time. I hope to have time in the coming week. I plan to find the test that makes that snapshot modification in a commit before the fix.

Initially, I believed the issue started at https://github.com/projen/projen/pull/3397, but when testing in `main` commits from the past few days, I found a different story.

From my tests, I think there are other causes besides #3397. I mean that I can reproduce today that the snapshot was outdated since the commits in main before the fix https://github.com/projen/projen/commit/75072c1ed699dd0e02ecd0627e72c4ff907c9728 . If tested today, multiple past commits will show the outdated snapshot `test/__snapshots__/new.test.ts.snap`. The snapshot validation looks correct, so when tested in the past, those past commits weren't modifying `test/__snapshots__/new.test.ts.snap`, and the GitHub workflow wasn't failing. That's why there should be an external factor not tracked in the projen git repository affecting the tests in `test/new.test.ts`.

It sounds unbelievable because the repository should have everything tracked in git, but my hypothesis is that something external to git exists in `test/new.test.ts`.

The tests consist in:

1. git checkout `git-ref-from-main-branch`
1. yarn && rm -rf lib && yarn projen && yarn test test/new.test.ts
1. git diff

These are the results of my tests
![image](https://github.com/projen/projen/assets/22875166/16f218bf-0476-42dc-a11a-92631e115986)

Here is an example of a test in v0.80.2
![image](https://github.com/projen/projen/assets/22875166/b87407fc-cf02-4de0-a6c6-574ce4dd7a98)

**Comment by giseburt** [comment]
> The snapshot mutation in CI exists so maintainers like me get the benefit of self-updating snapshots.

😵‍💫 I'm somewhat stunned by this statement, and even to a lesser extent this whole conversation. I appreciate everyone's time looking into this, but flaw isn't that it somehow didn't get committed. It's _that the tests passed!_

The snapshot is part of the test. It's a convenience feature so the test doesn't have to hold all that structure.

When the data being compared to the snapshot changes, the test is supposed to fail and show you that something that wasn't expected to change changed.

If you mutate that snapshot, you **change the test**. That's _designed_ to be a decision made by a person. 

When a test fails (snapshot or not), someone evaluates that failure and decides if it's the new code or the test (or both) that needs adjusted. It's legit to occasionally, carefully adjust a test, when the expectations of the program changed sufficiently to warrant it.

All that in mind, that there's even a "feature" that says "if a test fails, automatically change the test so it passes" completely blows me away.

🤯

**Comment by giseburt** [comment]
Ok, to make sure I had how snapshots are intended to be used right (and I'll happily admit when I have things wrong) I decided to dig around.

Here's what I found:

[Snapshot Testing with Jest](https://jestjs.io/docs/snapshot-testing):
   > A typical snapshot test case renders a UI component, takes a snapshot, then compares it to a reference snapshot file stored alongside the test. The test will fail if the two snapshots do not match: either the change is unexpected, or the reference snapshot needs to be updated to the new version of the UI component.
   > ...
   > 
   > The snapshot artifact should be committed alongside code changes, and reviewed as part of your code review process. (...) On subsequent test runs, Jest will compare the rendered output with the previous snapshot. If they match, the test will pass. If they don't match, either the test runner found a bug in your code (...) that should be fixed, or the implementation has changed and the snapshot needs to be updated.
   > ...
   > It's straightforward to spot when a snapshot test fails after a bug has been introduced. When that happens, go ahead and fix the issue and make sure your snapshot tests are passing again. Now, let's talk about the case when a snapshot test is failing due to an intentional implementation change.
   > ... Since we just updated our component to point to a different address, it's reasonable to expect changes in the snapshot for this component. Our snapshot test case is failing because the snapshot for our updated component no longer matches the snapshot artifact for this test case.
   > To resolve this, we will need to update our snapshot artifacts.
   >  …
   > **Best Practices : 1. Treat snapshots as code**
   > … The goal is to make it easy to review snapshots in pull requests, and fight against the habit of regenerating snapshots when test suites fail instead of examining the root causes of their failure.
   > …
   > **Are snapshots written automatically on Continuous Integration (CI) systems?**
   > It is expected that all snapshots are part of the code that is run on CI and since new snapshots automatically pass, they should not pass a test run on a CI system.

Also on that page are links to other blog posts and a video about using snapshots. 

Also on that page is notes about [Property Matchers](https://jestjs.io/docs/snapshot-testing#property-matchers) used with snapshots, which I feel are underutilized generally.

Sorry, I don’t want to come off as 🤓 “um, actually” but in this case this is a misused feature that can have a significant impact to the project, and more importantly is in a project that is designed to both help setup other projects and to act as a role-model (maybe somewhat unintentionally) to others, so I feel it’s important that this team specifically understand this.

**Comment by mrgrain** [comment]
@gmeligio I can confirm your observation. I went as far back as `ddfd7f32` and the snapshot now updates.

External things that can change:
- Node version
- yarn/npm version
- versions of dependencies used in the test case

**Comment by giseburt** [comment]
On quick clarification: I like the mutation-detection mechanism. I think it's great. I just don't think it's a substitute for the snapshot-match test failure.

**Comment by mrgrain** [comment]
> I understand and will comment on it no further. I don’t want to be a pain.

I'd love to discuss this further. I just don't want to distract from investigating this.

**Comment by giseburt** [comment]
On the topic at hand: “how did this not get detected?”

Here’s the exact location and time this could have been detected:

https://github.com/giseburt/projen/blob/dfa478baa4b511ce78ed8f5488d304ec62bc7c9f/test/new.test.ts#L50

The `execProjenCLI` call is defined here: https://github.com/giseburt/projen/blob/dfa478baa4b511ce78ed8f5488d304ec62bc7c9f/test/util.ts#L41

Which uses this:

```ts
const PROJEN_CLI = require.resolve("../lib/cli/index.js");
```

I can’t dig any further ATM, but my instinct tells me that since that’s in `../lib/` that means it’s compiled instead of executed directly (such as with `ts-node` like we do from the `projen` task in certain circumstances), which makes for an opportunity for it to be out of date.

🕵️‍♂️

**Comment by giseburt** [comment]
Couple other notes after reading others’ observations, my assumptions:

1. Change detection isn’t failing
   - It uses git, which we should be able to trust. The only likelihood I can think of is an ignore, but that’s super remote
2. The code in the commit does induce the change
   - The added [tests](https://github.com/giseburt/projen/commit/ada8e1260a26c70c0778a8f765eb3b0448092e2b) validate that independently

If those assumptions are correct, then the code being run in the test that should have had a snapshot change is not the code of the commit itself. I have no clue how that could be.

**Comment by gmeligio** [comment]
> Which uses this:
> 
> ```ts
> const PROJEN_CLI = require.resolve("../lib/cli/index.js");
> ```

I would say this part is good because it uses the source version controlled by git instead of an external version.

---- 

Today, I looked more into which tests modified the snapshots. By checking the tests in `test/new.test.ts`, disabling all at the beginning, and then starting to enable from the top, I found out the two tests that are causing changes `test/__snapshots__/new.test.ts.snap` are:

1. Test `projen new --from external`
https://github.com/projen/projen/blob/ac0edeeb3a0688d32369b4fc2aa5d17a3e940327/test/new.test.ts#L47-L71

2. Test `projen new --from external tarball`
https://github.com/projen/projen/blob/ac0edeeb3a0688d32369b4fc2aa5d17a3e940327/test/new.test.ts#L122-L149

I didn't see anything unusual on those two respect to the rest of the tests. There are other tests using **external** packages like `pepperize/projen-awscdk-app-ts` that didn't affect the snapshot. I need to debug further to see when the modification happens.

**Comment by giseburt** [comment]
Ok, I have traced it precisely. I'm not sure how to fix it. 😞  

TL;DR: 
<img width="1082" alt="image" src="https://github.com/projen/projen/assets/386001/8ac19ea7-024f-4c35-be0f-644cf4dc0e5d">

In these test, in spite of using `require.resolve("../lib/cli/index.js");` - an external `projen` binary is being used, the latest from NPM, it appears.

Ok, the trace:

1. In the test, we call `execProjenCLI`: `new.test.ts` → `test("projen new --from external", () => { ... }` → `execProjenCLI(projectdir, ...)`
2. Which calls `'node '<project_root>/lib/cli/index.js' new --from '@pepperize/projen-awscdk-app-ts@0.0.333' --no-post`
3. In (compiled form of) `index.ts` → `main()` → `const args = await ya.argv;` (a getter, `await` it likely wasted here)
4. (leaves projen code and disappears into the `yargs` code for 12 or so stacks)
5. Resurfaces in `src/cli/cmds/new.ts` in [`handler()`](https://github.com/projen/projen/blob/11fd842b54ad12a6bcce7eb6b6a12f029030a295/src/cli/cmds/new.ts#L114): `return await initProjectFromModule(process.cwd(), args.from, args);`
6. (skipping around a bit): `handler()` → `initProjectFromModule` → `initProject` → `Projects.createProject` → [`createProject`](https://github.com/projen/projen/blob/11fd842b54ad12a6bcce7eb6b6a12f029030a295/src/projects.ts#L157)
7. Calls `vm.runInContext(initProjectCode.join("\n"), ctx);` where `initProjectCode.join("\n"),` resolves to the following code:
    ```js
    const projectnk4pzlzjcye = new AwsCdkTypeScriptApp({
      __new__: {"args":{"cdkVersion":"2.1.0","defaultReleaseBranch":"main","devDeps":["@pepperize/projen-awscdk-app-ts@0.0.333"],"name":"my-project","projenrcTs":true},"fqn":"@pepperize/projen-awscdk-app-ts.AwsCdkTypeScriptApp","comments":"featured"},
      cdkVersion: "2.1.0",
      defaultReleaseBranch: "main",
      devDeps: ["@pepperize/projen-awscdk-app-ts@0.0.333"],
      name: "my-project",
      outdir: "/private/var/folders/11/h21t1_8s6t75ks54nkhwhj_40000gn/T/projen-test-ckCkKi/my-project",
      projenrcTs: true,

      // deps: [],                /* Runtime dependencies of this module. */
      // description: undefined,  /* The description is just a string that helps people understand the purpose of the package. */
      // packageName: undefined,  /* The "name" in package.json. */
    });
    projectnk4pzlzjcye.synth();
    ```
8. Which takes us to the compiled form of [@pepperize/projen-awscdk-app-ts/src/awscdk-app-ts.ts](https://github.com/pepperize/projen-awscdk-app-ts/blob/d367465fb53f4477bbf4b0a580b7fc168c53e95c/src/awscdk-app-ts.ts#L16) which includes the code `import { awscdk, javascript, Task } from "projen";` which, of course, resolves to the (newly installed from npm) `preen`, precise path at this time:
    `"/private/var/folders/11/h21t1_8s6t75ks54nkhwhj_40000gn/T/projen-test-ckCkKi/my-project/node_modules/projen/lib/index"`

In the specific case we're tracing here, I pushed a commit tot a PR that would fix `new --from` generating the wrong config (because it didn't detect the correct `jest` version) to the correct config, triggering the snapshot change. That commit was in a PR, NOT on npm (of course!), so the snapshot didn't change yet. It wasn't until a later PR that had enough time passed where the fix had propagated to npm, and triggered the snapshot change.

## Thoughts on a fix

The only solution I can think of would be to have the external project `package.json` adjusted with an [override](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#overrides) to a `file://` URL. Probably in `initProjectFromModule` where it "installs projen" (the problematic out-of-date one from npm, in fact).

**Comment by mrgrain** [comment]
Thanks for the detailed investigation! That's in line with what I expected to be going on.

I think in general `projen new` should always install the version of projen that the CLI is running on. Whether it's `projen@latest` or `projen@0.61.0` or a local package. What do you think?

Though I'm not sure yet how we can resolve that in a sensible way. 🤔

**Comment by giseburt** [comment]
> I think in general projen new should always install the version of projen that the CLI is running on. Whether it's projen@latest or projen@0.61.0 or a local package. What do you think?

I think the current mechanisms are fine outside of testing. For example, from nothing other than just installing node, `npx projen new ...` should install the version specified, with a reasonable default of `*`. Later on that same machine that same project is stall going to be cached by `npx` (unless you specify) and we don't want it to default to installing that one.

But inside testing we'd need to be able to specify a local file url for projen. That could be useful in development of external packages as well, I suppose. So if we made an option to specify the full projen dependency (like we do `--from`) to pass to npm, then we could use that in testing? And it could be used for unknown other reasons to specify a projen with other types of urls, like a github url. (Can you use that to point to a PR? I'd have to look, but I think you can. That could be useful.)

**Comment by mrgrain** [comment]
> Can you use that to point to a PR? I'd have to look, but I think you can. That could be useful.

I think npm can handle that, the problem is that with TypeScript (and jsii) you normally cannot use the repo directly since only TS files are checked-in, and not the produced output (i.e. JS files and the jsii assembly)

**Comment by gmeligio** [comment]
A question for picking the `projenVersion`: will it help for this issue if a mechanism like the one in `cdk init` is used?

https://github.com/aws/aws-cdk/blob/2378635ae278eb4c2f05ef1302c8737009022dc8/packages/aws-cdk/lib/init.ts#L161-L163
```typescript
const manifest = require(path.join(rootDir(), 'package.json'));
const MATCH_VER_BUILD = /\+[a-f0-9]+$/; // Matches "+BUILD" in "x.y.z-beta+BUILD"
const cdkVersion = manifest.version.replace(MATCH_VER_BUILD, '');
```

The `projen new` command supports `--projenVersion` (although not in the help) and having the current CLI version we could use `npx projen new --projenVersion <pickedVersion> typescript`.

**Comment by mrgrain** [comment]
So the question is if this `--projen-version` option can take a path to a local project.

**Comment by mrgrain** [comment]
`--projen-version file:${process.cwd()}` works. Unfortunately this needs to be sanitized in the output as well.

**Comment by giseburt** [comment]
I've been experimenting with that. I think we need two solutions:

1. `--projen-version file:${process.cwd()}` has the side-effects you mentioned for the snapshots, but instead adding the following to `execProjenCLI` right before the `exec(...)` seems to be much less disruptive, and only seems to need some `exclude: ['node_modules/**']` added where there wasn't a `node_modules` before:
    ```ts
    if (args.includes("--from")) {
      installPackage(workdir, `file://${process.cwd()}`, true); // fix for #3410
    }
    ```
2. `new XYZProject(...)` tests that call `projen` during the synth process need ``projenVersion:  `file://${process.cwd()}` `` (with the code adjusted to accept file urls) but then the `package.json` and `.projen/deps.json` need `s|"file://[^"]*"|"file:"//|g` (`sed` syntax, don't actually use `sed` 😄 ) run on them

**Comment by giseburt** [comment]
I am working on a PR for this BTW. I'm just getting bogged down in the day-to-day.

**Comment by mrgrain** [comment]
Thanks for the investigation. Pre-installing also sounds like a feasible option that might get us around the additional sanitization.

> 2\. `new XYZProject(...)` tests that call `projen` during the synth process need `` projenVersion:  `file://${process.cwd()}`  `` (with the code adjusted to accept file urls) but then the `package.json` and `.projen/deps.json` need `s|"file://[^"]*"|"file:"//|g` (`sed` syntax, don't actually use `sed` 😄 ) run on them

You got an example of those? This feels like an anti-pattern anyway.

**Comment by mrgrain** [comment]
Apologies @giseburt I was doing a refactor for `new.test.ts` already and ended up including your suggestion (at least the first part).

Let me know what you think: https://github.com/projen/projen/pull/3483

**Comment by giseburt** [comment]
@mrgrain NP. I was actually working on a different PR and was going to write tests and found tests that should fail/change snapshots were not, so having this fixed would be great. I’ll take a look asap.

**Comment by giseburt** [comment]
Btw I’m sorry I didn’t get to respond @mrgrain before it was closed but I didn’t really have anything to add. I’m glad this is sorted out.

**Comment by mrgrain** [comment]
No worries, I figured it's better than what we had before and we could always do a follow up. Thank you @gmeligio for helping with the investigation and thank you @giseburt for investigating and proposing the pre-install idea.
